File: backend/domain/segmentation.py
import re
from dataclasses import dataclass
from typing import Literal


@dataclass(frozen=True, slots=True)
class Segment:
    index: int
    start_line: int
    end_line: int
    prefix_lines: tuple[str, ...] = ()
    body_lines: tuple[str, ...] = ()
    suffix_lines: tuple[str, ...] = ()

    @property
    def text(self) -> str:
        return "\n".join((*self.prefix_lines, *self.body_lines, *self.suffix_lines))

    @property
    def body(self) -> str:
        return "\n".join(self.body_lines)


@dataclass(frozen=True, slots=True)
class Part:
    kind: Literal["separator", "segment"]
    lines: tuple[str, ...] = ()
    segment: Segment | None = None

    @property
    def is_segment(self) -> bool:
        return self.kind == "segment"


@dataclass(frozen=True, slots=True)
class AlignedPair:
    original: Segment | None = None
    translated: Segment | None = None


def _matches(line: str, patterns: tuple[str, ...]) -> bool:
    return any(re.search(pattern, line) for pattern in patterns)


def _split_delimiters(
    lines: list[str],
    start_patterns: tuple[str, ...],
    end_patterns: tuple[str, ...],
) -> tuple[tuple[str, ...], tuple[str, ...], tuple[str, ...]]:
    prefix = 0
    while prefix < len(lines) and _matches(lines[prefix], start_patterns):
        prefix += 1
    suffix = 0
    while suffix < len(lines) - prefix and _matches(lines[len(lines) - 1 - suffix], end_patterns):
        suffix += 1
    body_end = len(lines) - suffix if suffix else len(lines)
    return (
        tuple(lines[:prefix]),
        tuple(lines[prefix:body_end]),
        tuple(lines[body_end:]),
    )


def segment(
    text: str,
    start_patterns: tuple[str, ...] = (),
    end_patterns: tuple[str, ...] = (),
) -> tuple[Part, ...]:
    lines = text.split("\n")
    if not start_patterns:
        return (Part("separator", tuple(lines)),)

    total = len(lines)
    parts: list[Part] = []
    separator: list[str] = []
    index = 0
    i = 0
    while i < total:
        if _matches(lines[i], start_patterns):
            if separator:
                parts.append(Part("separator", tuple(separator)))
                separator = []
            end: int | None = None
            k = i
            while k < total:
                if k > i and _matches(lines[k], end_patterns):
                    end = k
                    break
                if k > i and _matches(lines[k], start_patterns):
                    end = k - 1
                    break
                k += 1
            if end is None:
                end = total - 1
            prefix, body, suffix = _split_delimiters(lines[i : end + 1], start_patterns, end_patterns)
            parts.append(
                Part(
                    "segment",
                    segment=Segment(index, i, end, prefix, body, suffix),
                )
            )
            index += 1
            i = end + 1
        else:
            separator.append(lines[i])
            i += 1

    if separator:
        parts.append(Part("separator", tuple(separator)))
    return tuple(parts)


def segment_by_separators(
    text: str,
    separator_patterns: tuple[str, ...],
) -> tuple[Part, ...]:
    lines = text.split("\n")
    if not separator_patterns:
        return (Part("separator", tuple(lines)),)

    parts: list[Part] = []
    buffer: list[str] = []
    marks: list[str] = []
    start = 0
    index = 0

    def flush_segment(end: int) -> None:
        nonlocal buffer, index
        if buffer:
            parts.append(
                Part("segment", segment=Segment(index, start, end, (), tuple(buffer), ()))
            )
            index += 1
            buffer = []

    def flush_marks() -> None:
        nonlocal marks
        if marks:
            parts.append(Part("separator", tuple(marks)))
            marks = []

    for position, line in enumerate(lines):
        if _matches(line, separator_patterns):
            flush_segment(position - 1)
            marks.append(line)
        else:
            flush_marks()
            if not buffer:
                start = position
            buffer.append(line)

    flush_segment(len(lines) - 1)
    flush_marks()
    return tuple(parts)


def reconstruct(parts: tuple[Part, ...]) -> str:
    lines: list[str] = []
    for part in parts:
        if part.kind == "segment" and part.segment is not None:
            lines.extend(part.segment.prefix_lines)
            lines.extend(part.segment.body_lines)
            lines.extend(part.segment.suffix_lines)
        else:
            lines.extend(part.lines)
    return "\n".join(lines)


def segments_of(parts: tuple[Part, ...]) -> tuple[Segment, ...]:
    return tuple(part.segment for part in parts if part.is_segment and part.segment)


def align(
    original_parts: tuple[Part, ...],
    translated_parts: tuple[Part, ...],
) -> tuple[AlignedPair, ...]:
    originals = segments_of(original_parts)
    translated = segments_of(translated_parts)
    count = max(len(originals), len(translated))
    return tuple(
        AlignedPair(
            original=originals[i] if i < len(originals) else None,
            translated=translated[i] if i < len(translated) else None,
        )
        for i in range(count)
    )

File: frontend/src/segmentation.ts
export interface Segment {
  index: number;
  startLine: number;
  endLine: number;
  prefixLines: string[];
  bodyLines: string[];
  suffixLines: string[];
}

export interface Part {
  kind: "separator" | "segment";
  lines: string[];
  segment?: Segment;
}

export interface AlignedPair {
  original?: Segment;
  translated?: Segment;
}

export function segmentText(segment: Segment): string {
  return [...segment.prefixLines, ...segment.bodyLines, ...segment.suffixLines].join("\n");
}

export function segmentBody(segment: Segment): string {
  return segment.bodyLines.join("\n");
}

function matches(line: string, patterns: string[]): boolean {
  return patterns.some((pattern) => new RegExp(pattern).test(line));
}

function splitDelimiters(
  lines: string[],
  startPatterns: string[],
  endPatterns: string[],
): { prefix: string[]; body: string[]; suffix: string[] } {
  let prefix = 0;
  while (prefix < lines.length && matches(lines[prefix], startPatterns)) {
    prefix += 1;
  }
  let suffix = 0;
  while (
    suffix < lines.length - prefix &&
    matches(lines[lines.length - 1 - suffix], endPatterns)
  ) {
    suffix += 1;
  }
  const bodyEnd = suffix > 0 ? lines.length - suffix : lines.length;
  return {
    prefix: lines.slice(0, prefix),
    body: lines.slice(prefix, bodyEnd),
    suffix: lines.slice(bodyEnd),
  };
}

export function segment(
  text: string,
  startPatterns: string[] = [],
  endPatterns: string[] = [],
): Part[] {
  const lines = text.split("\n");
  if (startPatterns.length === 0) {
    return [{ kind: "separator", lines: [...lines] }];
  }

  const total = lines.length;
  const parts: Part[] = [];
  let separator: string[] = [];
  let index = 0;
  let i = 0;

  while (i < total) {
    if (matches(lines[i], startPatterns)) {
      if (separator.length > 0) {
        parts.push({ kind: "separator", lines: separator });
        separator = [];
      }
      let end: number | undefined;
      let k = i;
      while (k < total) {
        if (k > i && matches(lines[k], endPatterns)) {
          end = k;
          break;
        }
        if (k > i && matches(lines[k], startPatterns)) {
          end = k - 1;
          break;
        }
        k += 1;
      }
      if (end === undefined) {
        end = total - 1;
      }
      const { prefix, body, suffix } = splitDelimiters(
        lines.slice(i, end + 1),
        startPatterns,
        endPatterns,
      );
      parts.push({
        kind: "segment",
        lines: [],
        segment: { index, startLine: i, endLine: end, prefixLines: prefix, bodyLines: body, suffixLines: suffix },
      });
      index += 1;
      i = end + 1;
    } else {
      separator.push(lines[i]);
      i += 1;
    }
  }

  if (separator.length > 0) {
    parts.push({ kind: "separator", lines: separator });
  }
  return parts;
}

export function reconstruct(parts: Part[]): string {
  const lines: string[] = [];
  for (const part of parts) {
    if (part.kind === "segment" && part.segment) {
      lines.push(
        ...part.segment.prefixLines,
        ...part.segment.bodyLines,
        ...part.segment.suffixLines,
      );
    } else {
      lines.push(...part.lines);
    }
  }
  return lines.join("\n");
}

export function segmentsOf(parts: Part[]): Segment[] {
  return parts
    .filter((part) => part.kind === "segment" && part.segment)
    .map((part) => part.segment as Segment);
}

export function align(originalParts: Part[], translatedParts: Part[]): AlignedPair[] {
  const originals = segmentsOf(originalParts);
  const translated = segmentsOf(translatedParts);
  const count = Math.max(originals.length, translated.length);
  const pairs: AlignedPair[] = [];
  for (let i = 0; i < count; i += 1) {
    pairs.push({ original: originals[i], translated: translated[i] });
  }
  return pairs;
}
